### ISSUE TITLE
saving space for a footer in GreedyBytes

### ISSUE THREAD

**Issue by girtsf** [construct/construct#701]
I'm looking for an easy way to say "this field fills the rest of the structure except for this fixed-size footer". I currently have a hack with something like this:

```
    [..]
    'data' / FocusedSeq(1,
        'remainder' / Rebuild(Peek(GreedyBytes), this.data + b'  '),
        'data' / Bytes(len_(this.remainder) - 2),
    ),
```

This seems to build and parse correctly, but is a bit... ugly. Is there a better way?

**Comment by arekbulski** [comment]
I dont understand the Rebuild/Peek combo. Peek does not build (noop). Rebuild is used to compute the data to be used for building. Thats like cooking but not eating. 

I dont understand what it is that you are trying to achieve either.

**Comment by girtsf** [comment]
Sorry for the confusing example. Maybe let me rephrase - let's say I have a format like this:
* field "a", one byte integer
* field "b", bytes, that takes up all the space between "a" and "c"
* field "c", one byte integer

What's the right way to describe this with Construct?

**Comment by arekbulski** [comment]
Is the middle field of known size, or its EOF -1 byte?
```py
Struct(
  'a' / Byte,
  Padding(???-1),
  'c' / Byte,
)
```

**Comment by arekbulski** [comment]
```py
FocusedSeq('ac'
    'entire' / GreedyBytes,
    'ac' / RestreamData(this.entire, Struct(
        'a' / Byte,
        Padding(len_(this.entire)-1),
        'c' / Byte,
    )),
)

```

**Comment by arekbulski** [comment]
This one fails (and I am opening a bug report for it):
```py
>>> d = FocusedSeq('ac'
...     'entire' / GreedyBytes,
...     'ac' / RestreamData(this.entire, Struct(
...         'a' / Byte,
...         Padding(len_(this._.entire)-1),
...         'c' / Byte,
...     )),
... )
>>> print(d.parse(b'\x01GGGGGGGGGG\x02'))
KeyError: 'entire'
```
But this one works fine:
```py
>>> d = Struct(
...     'entire' / GreedyBytes,
...     Seek(0),
...     'a' / Byte,
...     Seek(-1, 2),
...     'c' / Byte,
... )
>>> print(d.parse(b'\x01GGGGGGGGGG\x02'))
Container: 
    entire = b'\x01GGGGGGGGGG\x02' (total 12)
    a = 1
    c = 2
```

**Comment by girtsf** [comment]
Oh, `RestreamData` and `Seek(-1, 2)` are both nifty.

Though in my case I also needed to be able to extract field `b`, as well as `build` it.

**Comment by arekbulski** [comment]
It might be easier to have 2 different constructs for parsing and building. You could use:
```py
IfThenElse(this._isparsing, StructForParsing, StructForBuilding)
```
I would need to change parse/build methods to add _isparsing _isbuilding context entries.

**Comment by arekbulski** [comment]
Voila:
```py
>>> d = Struct(
...     'data' / Rebuild(Peek(GreedyBytes), this.b + b'\x01\x02'),
...     'a' / Byte,
...     'b' / Bytes(len_(this.data)-2),
...     'c' / Byte,
... )
>>> print(d.parse(b'\x01GGG\x02'))
Container: 
    data = b'\x01GGG\x02' (total 5)
    a = 1
    b = b'GGG' (total 3)
    c = 2
>>> print(d.build(dict(a=1, b=b'GGG', c=2)))
b'\x01GGG\x02'
```

**Comment by girtsf** [comment]
> Voila:
> `...     'data' / Rebuild(Peek(GreedyBytes), this.b + b'\x01\x02'),`

And that was my Rebuild/Peek combo from earlier. :)

**Comment by arekbulski** [comment]
Ha ha, right! Except you also managed to explain to me what is its purpose. I am even considering puting this code into the docs tutorial.

**Comment by arekbulski** [comment]
You might also use this one, at least for parsing. Its a very old (classic) way of measuring file size on Linux, because Linux does not have any syscalls for file size, you measure it by seeking to EOF and measuring current offset. Well, you can do something very similar in Construct:
```py
d = Struct(
    'a' / Byte,
    'offset1' / Tell,
    Seek(-1, 2),
    'offset2' / Tell,
    'b' / Bytes(this.offset2-this.offset1),
    'c' / Byte,
)
```

**Comment by arekbulski** [comment]
I am also considering adding a new class: GreedyBytesEx(startoffset=0, endoffset=0). With default parameters it would behave just like GreedyBytes but it would allow you to parse till EOF-N, in your case EOF-1. Then you could simply use a parsing/building scheme, very simple one:
```py
d = Struct(
    'a' / Byte,
    'b' / GreedyBytesEx(endoffset=-1),
    'c' / Byte,
)
```
Would you be interested? I can implement it for you.

**Comment by girtsf** [comment]
I do have a workaround that works for me for now, but something like `GreedyBytesEx` would work well for my use case.

**Comment by arekbulski** [comment]
I will gladly add GreedyBytesEx for you, but I have a backlog of features to implement (and rather high priority at that) so please give me some time to get to it. I am treating your case as high priority.

**Comment by timrid** [comment]
Are there any plans to implement GreedyBytesEx?
